src/exo/LoopIR_scheduling.py

```diff
@@ -2564,11 +2564,147 @@ def map_e(self, e):
         return e
 
 
+
+class _DoNormalize(LoopIR_Rewrite):
+    def __init__(self, proc):
+        self.C = Sym("temporary_constant_symbol")
+        super().__init__(proc)
+
+        self.proc = InferEffects(self.proc).result()
+
+    def concat_map(self, op, lhs, rhs):
+        if op == '+':
+            # if has same key: add value
+            common = { key: (lhs[key]+rhs[key]) for key in lhs if key in rhs }
+            return lhs | rhs | common
+        elif op == '-':
+            # has same key: sub value
+            common = { key: (lhs[key]-rhs[key]) for key in lhs if key in rhs }
+            # else, negate the rhs and cat map
+            neg_rhs = { key: -rhs[key] for key in rhs }
+            return lhs | neg_rhs | common
+        elif op == '*':
+            # rhs or lhs NEEDS to be constant
+            assert len(rhs) == 1 or len(lhs) == 1
+            # multiply the other one's value by that constant
+            if len(rhs) == 1 and self.C in rhs:
+                return { key: lhs[key]*rhs[self.C] for key in lhs }
+            else:
+                assert self.C in lhs
+                return { key: rhs[key]*lhs[self.C] for key in rhs }
+        else:
+            assert False, "bad case"
+
+    def normalize_e(self, e):
+        if isinstance(e, LoopIR.Read):
+            assert e.type.is_indexable()
```

Comment:
Should this assertion be lifted up to the top scope of this method, applicable to all cases?